### PR TITLE
Implement home cleanup utility

### PR DESCRIPTION
**Pull request by triagonal**
### Information

This PR implements a home cleanup utility subcommand inside the `/essentials` command, similar to the existing `/ess cleanup` subcommand for cleaning up user files. A "home deleter" is something that's requested fairly often on the MOSS support server, primarily from server owners looking to do world resets. This also works as a solution for #4088.

**Environments tested:**

OS: Windows 10 20H2
Java version: `openjdk version "11.0.10"`

- [x] 1.16.5-R0.1-SNAPSHOT git-Paper-777 (MC: 1.16.5)

### CODE REVIEW

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 1]
```diff
@@ -253,7 +260,78 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                            continue;
+                        }
+                        for (String homeName : user.getHomes()) {
+                            try {
+                                final Location home = user.getHome(homeName);
+                                if (home != null && home.getWorld() != null && home.getWorld().getName().equals(worldName)) {
+                                    user.delHome(homeName);
+                                }
+                            } catch (Exception e) {
+                                ess.getLogger().info("Unable to delete home " + homeName + " for " + user.getName());
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:315]:
```suggestion
                                ess.getLogger().warning("Unable to delete home " + homeName + " for " + user.getName());
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 2]
```diff
@@ -253,7 +260,78 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                            }
+                        }
+                    }
+                    sender.sendMessage(tl("cleanedWorldHomes", worldName));
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]: Would be useful to also notify the executor if any error occurred

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 3]
```diff
@@ -253,7 +260,78 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                });
+                break;
+            case "world":
+                if (args.length < 3) {
+                    sender.sendMessage("Provide a world name to delete homes inside a specific world.");
+                    throw new Exception("/<command> homecleanup (invalid | world <world>)");
+                }
+
+                final String worldName = args[2];
+                if (Bukkit.getWorld(worldName) == null) {
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]:
```suggestion
                if (server.getWorld(worldName) == null) {
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 4]
```diff
@@ -438,6 +517,13 @@ private void runVersion(final Server server, final CommandSource sender, final S
                     return Lists.newArrayList("-1", "0");
                 }
                 break;
+            case "homecleanup":
+                if (args.length == 2) {
+                    return Lists.newArrayList("invalid", "world");
+                } else if (args.length == 3 && args[1].equalsIgnoreCase("world")) {
+                    return Bukkit.getWorlds().stream().map(World::getName).collect(Collectors.toList());
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]:
```suggestion
                    return server.getWorlds().stream().map(World::getName).collect(Collectors.toList());
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 5]
```diff
@@ -12,9 +12,12 @@
 import com.google.common.base.Charsets;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
+import org.bukkit.Bukkit;
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]:
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 6]
```diff
@@ -253,7 +261,71 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
         });
+    }
+
+    private void runHomes(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
+        if (args.length < 2) {
+            sender.sendMessage("This sub-command provides a utility to mass-delete homes based on user options:");
+            sender.sendMessage("Use \"fix\" to delete all homes inside non-existent or unloaded worlds.");
+            sender.sendMessage("Use \"delete\" to delete all existing homes");
+            sender.sendMessage("Use \"delete <worldname>\" to delete all homes inside a specific world.");
+            throw new Exception(HOMES_USAGE);
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:270]:
```suggestion
            throw new NotEnoughArgumentsException(HOMES_USAGE);
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 7]
```diff
@@ -253,7 +261,71 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                            }
+                        }
+                    }
+                    sender.sendMessage(tl("fixedHomes"));
+                });
+                break;
+            case "delete":
+                final boolean filterByWorld = args.length >= 3;
+                if (filterByWorld && server.getWorld(args[2]) == null) {
+                    sender.sendMessage("The world \"" + args[2] + "\" does not exist.");
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]: Translations

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 8]
```diff
@@ -253,7 +261,71 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                            } catch (Exception e) {
+                                ess.getLogger().info("Unable to delete home " + homeName + " for " + user.getName());
+                            }
+                        }
+                    }
+                    sender.sendMessage(filterByWorld ? tl("deletedHomesWorld", args[2]) : tl("deletedHomes"));
+                });
+                break;
+            default:
+                sender.sendMessage("Invalid arguments.");
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]: translation

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 9]
```diff
@@ -253,7 +261,71 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                        }
+                    }
+                    sender.sendMessage(tl("fixedHomes"));
+                });
+                break;
+            case "delete":
+                final boolean filterByWorld = args.length >= 3;
+                if (filterByWorld && server.getWorld(args[2]) == null) {
+                    sender.sendMessage("The world \"" + args[2] + "\" does not exist.");
+                    throw new Exception(HOMES_USAGE);
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]:
```suggestion
                    throw new NotEnoughArgumentsException(HOMES_USAGE);
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 10]
```diff
@@ -253,7 +261,71 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                                ess.getLogger().info("Unable to delete home " + homeName + " for " + user.getName());
+                            }
+                        }
+                    }
+                    sender.sendMessage(filterByWorld ? tl("deletedHomesWorld", args[2]) : tl("deletedHomes"));
+                });
+                break;
+            default:
+                sender.sendMessage("Invalid arguments.");
+                throw new Exception(HOMES_USAGE);
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:323]:
```suggestion
                throw new NotEnoughArgumentsException(HOMES_USAGE);
```

**Review by JRoy — DISMISSED** [review]

**Review by JRoy — APPROVED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 11]
```diff
@@ -38,6 +41,8 @@
     private static final Sound NOTE_HARP = EnumUtil.valueOf(Sound.class, "BLOCK_NOTE_BLOCK_HARP", "BLOCK_NOTE_HARP", "NOTE_PIANO");
     private static final Sound MOO_SOUND = EnumUtil.valueOf(Sound.class, "COW_IDLE", "ENTITY_COW_MILK");
 
+    private static final String HOMES_USAGE = "/<command> homes (fix | delete [world])";
+
```
Comment (l1ttleO) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:45]:
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 12]
```diff
@@ -253,7 +259,69 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
+                            } catch (Exception e) {
+                                ess.getLogger().info("Unable to delete home " + homeName + " for " + user.getName());
+                            }
+                        }
+                    }
+                    sender.sendMessage(filterByWorld ? tl("deletedHomesWorld", args[2]) : tl("deletedHomes"));
+                });
+                break;
+            default:
+                throw new Exception(HOMES_USAGE);
```
Comment (l1ttleO) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:323]:
```suggestion
                throw new Exception("/<command> homes (fix | delete [world])");
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 13]
```diff
@@ -253,7 +259,69 @@ private void runCleanup(final Server server, final CommandSource sender, final S
...
         });
+    }
 
+    private void runHomes(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
+        if (args.length < 2) {
+            sender.sendMessage("This sub-command provides a utility to mass-delete homes based on user options:");
+            sender.sendMessage("Use \"fix\" to delete all homes inside non-existent or unloaded worlds.");
+            sender.sendMessage("Use \"delete\" to delete all existing homes.");
+            sender.sendMessage("Use \"delete <worldname>\" to delete all homes inside a specific world.");
+            throw new Exception(HOMES_USAGE);
```
Comment (l1ttleO) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:270]:
```suggestion
            throw new Exception("/<command> homes (fix | delete [world])");
```
this is for consistency

**Comment by LucidAPs** [comment]
This is a great tool! How about also making a command that would delete broken uuid files from userdata and usermaps with the npc flag, duplicates etc.

**Pull request merged**